src/window/mod.rs

```diff
@@ -0,0 +1,170 @@
+        C: PixelColor + Into<Rgb888> + From<Rgb888>,
+    {
+        let start = Instant::now();
+
+        if let Ok(path) = env::var("EG_SIMULATOR_CHECK") {
+            let output = display.to_rgb_output_image(&self.output_settings);
+
+            let png_file = BufReader::new(File::open(path).unwrap());
+            let expected = image::load(png_file, image::ImageFormat::Png)
+                .unwrap()
+                .to_rgb8();
+
+            let png_size = Size::new(expected.width(), expected.height());
+
+            assert!(
+                output.size().eq(&png_size),
+                "display dimensions don't match PNG dimensions (display: {}x{}, PNG: {}x{})",
+                output.size().width,
+                output.size().height,
+                png_size.width,
+                png_size.height
+            );
+
+            assert!(
+                output
+                    .as_image_buffer()
+                    .as_raw()
+                    .eq(&expected.as_raw().deref()),
+                "display content doesn't match PNG file",
+            );
+
+            process::exit(0);
+        }
+
+        if let Ok(path) = env::var("EG_SIMULATOR_CHECK_RAW") {
+            let expected = SimulatorDisplay::load_png(path).unwrap();
+
+            assert!(
+                display.size().eq(&expected.size()),
+                "display dimensions don't match PNG dimensions (display: {}x{}, PNG: {}x{})",
+                display.size().width,
+                display.size().height,
+                expected.size().width,
+                expected.size().height
+            );
+
+            assert!(
+                display.pixels.eq(&expected.pixels),
+                "display content doesn't match PNG file",
+            );
+
+            process::exit(0);
+        }
+
+        if let Ok(path) = env::var("EG_SIMULATOR_DUMP") {
+            display
+                .to_rgb_output_image(&self.output_settings)
+                .save_png(path)
+                .unwrap();
+            process::exit(0);
+        }
+
+        if let Ok(path) = env::var("EG_SIMULATOR_DUMP_RAW") {
+            display
+                .to_rgb_output_image(&OutputSettings::default())
+                .save_png(path)
+                .unwrap();
+            process::exit(0);
+        }
+
+        #[cfg(feature = "with-sdl")]
+        {
+            if self.framebuffer.is_none() {
+                self.framebuffer = Some(OutputImage::new(display, &self.output_settings));
+            }
+
+            if self.sdl_window.is_none() {
+                self.sdl_window = Some(SdlWindow::new(display, &self.title, &self.output_settings));
+            }
+
+            let framebuffer = self.framebuffer.as_mut().unwrap();
+            let sdl_window = self.sdl_window.as_mut().unwrap();
+
+            framebuffer.update(display);
+            sdl_window.update(&framebuffer);
+        }
+
+        thread::sleep(start + self.desired_loop_duration - Instant::now());
```

Comment:
This sleep assumes my processing takes 0 time. Can you store the time after the thread wakes up, then use that as a reference instead of grabbing it at the start of `update`?

Reply:
Thanks for noticing this! I'll apply these suggestions and add the section in `CHANGELOG.md` that @jamwaffles suggested this afternoon. 👍🏽

Reply:
```rust
thread::sleep(
    (self.frame_start + self.desired_loop_duration)
        .saturating_duration_since(Instant::now()),
);

self.frame_start = Instant::now();
```

If your frame takes a long time and becomes "negative", `saturating_duration_since` will return zero, which I think will fix the issue.

Reply:
Thanks for decoding my awkward comment perfectly :)